File: leetcode3/이진희/3912_Valid Elements in an Array.java
/*

1. 아이디어 : 왼쪽과 오른쪽을 각각 계산하여, 현재 값이 모든 왼쪽 값보다 큰지, 현재 값이 모든 오른쪽 값보다 큰지 체크
            boolean 배열에 표시해 뒀다가 마지막에 계산 

2. 시간복잡도 : O(3*N)

3. 자료구조/알고리즘 : 완전탐색

 */

class Solution {
    public List<Integer> findValidElements(int[] nums) {

        List<Integer> ans = new ArrayList<>();
        boolean[] check = new boolean[nums.length];

        check[0] = true;
        check[nums.length-1] = true;
        
        int leftMax = nums[0];
        for(int i=1; i<nums.length; i++){
            if(nums[i] > leftMax) check[i] = true;
            leftMax = Math.max(leftMax, nums[i]);
        }

        int rightMax = nums[nums.length-1];
        for(int i=nums.length-2; i>=0; i--){
            if(nums[i] > rightMax) check[i] = true;
            rightMax = Math.max(rightMax, nums[i]);
        }

        for(int i=0; i<nums.length; i++){
            if(check[i]) ans.add(nums[i]);
        }

        return ans;
    }
}

File: leetcode3/최원준/3912. Valid Elements in an Array.py
#

'''
1. 아이디어 :
왼쪽부터 오른쪽(left_max), 오른쪽부터 왼쪽(right_max)으로 가면서 만나는 숫자들의 최대값을 저장하는 배열을 만듭니다.
nums를 순회하면서 left_max[i]보다 크거나 right_max[i]보다 크면 후보에 포함합니다.

2. 시간복잡도 :
    O(n)

3. 자료구조/알고리즘 :
-

'''


class Solution:
    def findValidElements(self, nums: list[int]) -> list[int]:
        n = len(nums)
        left_max = [0] * n
        right_max = [0] * n

        for i in range(n-1):
            left_max[i+1] = max(left_max[i], nums[i])
            right_max[n-i-2] = max(right_max[n-i-1], nums[n-i-1])

        return [nums[i] for i in range(n) if nums[i] > left_max[i] or nums[i] > right_max[i]]